nomt/src/rollback/mod.rs

```diff
@@ -0,0 +1,342 @@
+//!
+//! The deltas are also persisted on disk in a [`seglog`].
+
+use std::{
+    collections::{BTreeMap, VecDeque},
+    fs::File,
+    io::Cursor,
+    path::PathBuf,
+    sync::Arc,
+};
+
+use dashmap::DashMap;
+use nomt_core::trie::KeyPath;
+use parking_lot::Mutex;
+use threadpool::ThreadPool;
+
+use self::delta::Delta;
+use crate::{
+    seglog::{self, RecordId, SegmentedLog},
+    KeyReadWrite,
+};
+
+mod delta;
+#[cfg(test)]
+mod tests;
+
+const MAX_SEGMENT_SIZE: u64 = 64 * 1024 * 1024; // 64 MiB
+
+struct InMemory {
+    /// The log of deltas that we have accumulated so far modulo the [`Self::head`].
+    ///
+    /// The items are pushed onto the back and popped from the front. When the log reaches
+    /// `limit_log_len`, the oldest delta is discarded.
+    ///
+    /// The deltas are stored in-memory even after they are dumped on disk. Upon restart, the deltas
+    /// are re-read from disk and stored here.
+    log: VecDeque<(RecordId, Delta)>,
+
+    /// If this is set, then the next writeout will truncate the log at this offset.
+    pending_truncate: Option<u64>,
+}
+
+struct Shared {
+    worker_tp: ThreadPool,
+    in_memory: Mutex<InMemory>,
+    seglog: Mutex<SegmentedLog>,
+    /// The number of items that we should keep in the log. Deltas that are past this limit are
+    /// discarded.
+    max_rollback_log_len: usize,
+}
+
+impl InMemory {
+    fn new() -> Self {
+        Self {
+            log: VecDeque::new(),
+            pending_truncate: None,
+        }
+    }
+
+    /// Push a delta into the in-memory cache.
+    fn push_back(&mut self, record_id: RecordId, delta: Delta) {
+        self.log.push_back((record_id, delta));
+    }
+
+    fn pop_back(&mut self) -> Option<(RecordId, Delta)> {
+        self.log.pop_back()
+    }
+
+    // Returns the total number of deltas, including the staged one.
+    fn total_len(&self) -> usize {
+        self.log.len()
+    }
+}
+
+/// This structure manages the rollback log. Modifications to the rollback log are made using
+/// [`RollbackSnapshotBuilder`].
+#[derive(Clone)]
+pub struct Rollback {
+    shared: Arc<Shared>,
+}
+
+impl Rollback {
+    pub fn read(
+        max_rollback_log_len: u32,
+        db_dir_path: PathBuf,
+        db_dir_fd: File,
+        rollback_start_active: u64,
+        rollback_end_active: u64,
+    ) -> anyhow::Result<Self> {
+        let mut in_memory = InMemory::new();
+        let seglog = seglog::open(
+            db_dir_path,
+            db_dir_fd,
+            "rollback".to_string(),
+            MAX_SEGMENT_SIZE,
+            rollback_start_active.into(),
+            rollback_end_active.into(),
+            |record_id, payload| {
+                let mut cursor = Cursor::new(payload);
+                let delta = Delta::decode(&mut cursor)?;
+                in_memory.push_back(record_id, delta);
+                Ok(())
+            },
+        )?;
+        let shared = Arc::new(Shared {
+            worker_tp: ThreadPool::new(4),
```

Comment:
should this be parameterized somehow in `Options`?

Reply:
It's not clear this needs to be a thread-pool, either - one thread is usually sufficient for saturating I/O. But I guess that'd require an asynchronous version of `load_value`.

Reply:
Let's file a follow-up for that